src/combinatorics/simple_graph/metric.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Vincent Beffara
+-/
+import combinatorics.simple_graph.connectivity
+import data.nat.lattice
+
+/-!
+# Graph metric
+
+This module defines the `simple_graph.dist` function, which takes
+pairs of vertices to the length of the shortest walk between them.
+
+## Main definitions
+
+- `simple_graph.dist` is the graph metric.
+
+## Todo
+
+- Provide an additional computable version of `simple_graph.dist`
+  for when `G` is connected.
+
+## Tags
+
+graph metric
+
+-/
+
+namespace simple_graph
+variables {V : Type*} (G : simple_graph V)
+
+/-! ## Metric -/
+
+/-- The distance between two vertices is the length of the shortest walk between them.
+If no such walk exists, this uses the junk value of `0`. -/
+noncomputable
+def dist (u v : V) : ℕ := Inf (set.range (walk.length : G.walk u v → ℕ))
```

Comment:
Perhaps this isn't an improvement, but can we instead use `nat.find` composed with an if/then/else? My guess is that would be easier to work with, but perhaps you've already tried it!

Reply:
My first iteration was `nat.find` taking the hypothesis that `u` and `v` are connected, in any case the `dist` function would mostly be used under this hypothesis. My preference would be for this plus an `enat` version to account for the other case when `u` and `v` are not connected. I guess having `Inf` and a junk value is the usual tradeoff, but as a total beginner I defer this kind of design choices to others :-)

Reply:
@b-mehta My thought with `Inf` vs `nat.find` was that it's better for definitions to be the "mathematical" one if possible. I also like how this lets us use general properties of infima. (What were you suggesting with if/then/else? Was it to essentially [unfold the definition](https://github.com/leanprover-community/mathlib/blob/92e675977762b0183d67572ca4db23314c57bd88/src/data/nat/lattice.lean#L26)?) If we're worrying about computability, there are better algorithms than `nat.find` for finding shortest distances in graphs -- I'd rather we define these separately and prove they compute the mathematical definition.

@vbeffara There's a heuristic that definitions that take fewer hypotheses tend to be easier to use. A reason we don't need a connectivity hypothesis in `dist` is because the only function it would serve is to act as a token certifying it's not being used for graphs you don't want it to be used for -- but think about how theorems about `dist` would have the connectivity hypothesis on hand anyway, so there's not much to be gained by requiring that this information be passed along to `dist` itself.

Regarding `enat`, I'm gambling on the common case for `dist` being connected graphs, where `enat` would add complexity over plain `nat`s. It would probably be less painful having both `dist` and an `edist` than trying to always work with the "right" codomain. Plus, you can easily check when `dist` has given you a junk value: it's exactly when `u` is not equal to `v` and `dist u v = 0`. None of this is a proof that this is the right way to go, and unfortunately we're lacking in theorems to test this all out.

Reply:
I agree that computability shouldn't be a concern here. My idea was to unfold the definition but also to simplify the mem-set-range in that hypothesis - but I'm happy either way round!